core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java

```diff
@@ -132,6 +132,37 @@ public UfsStatus getStatus(AlluxioURI path) {
     return mStatuses.get(path);
   }
 
+  /**
+   * Attempts to return a status from the cache. If it doesn't exist, reaches to the UFS for it.
+   *
+   * @param path the path the retrieve
+   * @param mountTable the Alluxio mount table
+   * @return The corresponding {@link UfsStatus} or {@code null} if there is none stored
+   */
+  @Nullable
+  public UfsStatus fetchStatusIfAbsent(AlluxioURI path, MountTable mountTable)
+      throws InvalidPathException {
+    UfsStatus status = mStatuses.get(path);
+    if (status != null) {
+      return status;
+    }
```

Comment:
you might want to use `computeIfAbsent` or `putIfAbsent` to implement this?  otherwise k concurrent calls to this function with the same path might result in k ufs calls  Using one of those methods ensures only one call to the UFS is made.

Reply:
Unfortunately I can't with the current implementation or else it will result in deadlock. This function calls `addStatus` which actually does the insert into the map. Calling `ConcurrentHashMap#put` while within a `computeIfAbsent` function will deadlock

Reply:
instead of calling `addStatus` perhaps you could just return the right value, which will result in the values being inserted into the map by `computeIfAbsent` right?

Reply:
I use `addStatus` because it performs a check to make sure the status has the correct naming scheme.

Reply:
I guess I can refactor out the check and put it elsewhere. It just seems more error prone, because then statuses could still be inserted without that check performed

Reply:
i see, sounds good to me